### ISSUE TITLE
--reset is crashing on when sync is called

### ISSUE THREAD

**Issue by tcezard** [EdinburghGenomics/Analysis-Driver#153]
I pull master in branch laneduprate and wanted to run some test.
but when I reset a run on ultra it crashed with the following error

``` python
Traceback (most recent call last):
  File "/home/U008/tcezard/Applications/Analysis-Driver/dev/bin/edingen_analysis_driver.py", line 10, in <module>
    sys.exit(analysis_driver.client.main())
  File "/home/U008/tcezard/Applications/Analysis-Driver/dev/analysis_driver/client.py", line 38, in main
    scanner.get_dataset(d).reset()
  File "/home/U008/tcezard/Applications/Analysis-Driver/dev/analysis_driver/dataset.py", line 57, in reset
    self.most_recent_proc.change_status(DATASET_REPROCESS)
  File "/home/U008/tcezard/Applications/Analysis-Driver/dev/analysis_driver/dataset.py", line 237, in change_status
    self.update_entity(status=status)
  File "/home/U008/tcezard/Applications/Analysis-Driver/dev/analysis_driver/dataset.py", line 234, in update_entity
    self.sync()
  File "/home/U008/tcezard/Applications/Analysis-Driver/dev/analysis_driver/dataset.py", line 230, in sync
    raise RestCommunicationError('Sync failed: ' + str(patch_content))
analysis_driver.exceptions.RestCommunicationError: Sync failed: {'_links': {'self': {'title': 'analysis_driver_proc', 'href': 'analysis_driver_procs/run_150723_E00306_0025_BHCHK3CCXX_25_03_2016_15:01:03'}}, 'end_date': '25_03_2016_18:13:54', 'pid': 0, '_id': '56f552af3df5b1fb11da1dc7', '_etag': 'e5f3ccb9e59a2f498880e78526d3930ecc31fd6f', '_updated': '25_03_2016_18:13:54', 'status': 'reprocess', 'dataset_type': 'run', '_created': '25_03_2016_15:01:03', 'dataset_name': '150723_E00306_0025_BHCHK3CCXX', 'stages': [{'exit_status': 0, 'date_started': '25_03_2016_15:01:03', 'date_finished': '25_03_2016_15:03:19', 'stage_name': 'transfer'}, {'exit_status': 0, 'date_started': '25_03_2016_15:03:19', 'date_finished': '25_03_2016_15:03:19', 'stage_name': 'setup'}, {'exit_status': 0, 'date_started': '25_03_2016_15:03:20', 'date_finished': '25_03_2016_15:08:50', 'stage_name': 'bcl2fastq'}, {'exit_status': 0, 'date_started': '25_03_2016_15:08:50', 'date_finished': '25_03_2016_15:23:35', 'stage_name': 'sickle_filter'}, {'exit_status': 0, 'date_started': '25_03_2016_15:23:35', 'date_finished': '25_03_2016_15:30:18', 'stage_name': 'fastqc'}, {'exit_status': 0, 'date_started': '25_03_2016_15:23:35', 'date_finished': '25_03_2016_15:49:42', 'stage_name': 'seqtk_fqchk'}, {'exit_status': 0, 'date_started': '25_03_2016_15:23:35', 'date_finished': '25_03_2016_15:57:37', 'stage_name': 'md5sum'}, {'exit_status': 0, 'date_started': '25_03_2016_15:57:43', 'date_finished': '25_03_2016_18:13:54', 'stage_name': 'data_transfer'}]}
```

looking into it a bit more it seems that the PATCH failed due to unexpected fields in the most_recent_proc

``` python
{"_status": "ERR", "_issues": {
"_updated": "unknown field",
 "_created": "unknown field", 
"_etag": "unknown field", 
"_id": "unknown field", 
"_links": "unknown field"
}}. Status code 422. Reason: UNPROCESSABLE ENTITY
```

Could you look into it?

**Comment by tcezard** [comment]
I think f4cb9ad should fix this issue however it seems that the the stage collisions that used be silent is now making the process fail.
If I'm right we need to either:
- Make it fail silently again
- Fix EdinburghGenomics/Reporting-App#21

**Comment by tcezard** [comment]
Scrap the last comment, making the MostRecentProc object threadsafe did the trick.
see dbfacfed7e9a2a25e8e0ed00751a743a063fadf1
